### PR TITLE
Hash calculation using json serializer

### PR DESCRIPTION
**Pull request by marcinwadon**

### CODE REVIEW

File: modules/currency-l0/src/main/scala/org/tessellation/currency/l0/modules/Programs.scala [hunk 1]
```diff
@@ -22,7 +23,7 @@ import org.tessellation.security.{Hasher, SecurityProvider}
 
 object Programs {
 
-  def make[F[_]: Async: Random: KryoSerializer: Hasher: SecurityProvider](
+  def make[F[_]: Async: Random: KryoSerializer: JsonHashSerializer: Hasher: SecurityProvider](
```
Comment (marcinwadon) [inline modules/currency-l0/src/main/scala/org/tessellation/currency/l0/modules/Programs.scala:26]: I think a better name is needed, as it's not only for Hash now

Comment (scasplte2) [reply modules/currency-l0/src/main/scala/org/tessellation/currency/l0/modules/Programs.scala:26]: In my mind, "serialization" implies binary as one side of the Encode/Decode functionality. Perhaps `JsonSerializer` is enough to imply the intended behavior.

**Review by kpudlik — DISMISSED** [review]

File: modules/shared/src/main/scala/org/tessellation/security/signature/Signed.scala [hunk 2]
```diff
@@ -106,21 +106,31 @@ object Signed {
...
+            },
+          implicit encoder =>
+            Hasher[F]
+              .hashJson(signed.value)
+              .flatMap {
+                signature.verifySignatureProof(_, proof).flatMap {
+                  if (_) true.pure[F]
+                  else
+                    Hasher[F].hashKryo(signed.value) >>= { signature.verifySignatureProof(_, proof) }
+                }
```
[the remark this answers is not in the document]
Comment (marcinwadon) [reply modules/shared/src/main/scala/org/tessellation/security/signature/Signed.scala]: done :+1:

File: modules/currency-l0/src/main/scala/org/tessellation/currency/l0/cli/method.scala [hunk 3]
```diff
@@ -172,8 +179,44 @@ object method {
...
+              snapshot,
+              seedlistPath,
+              prioritySeedlistPath,
+              collateralAmount,
+              globalL0Peer,
+              l0TokenIdentifier,
+              trustRatingsPath,
+              lastKryoHash
+            ) =>
+          val lastKH =
```
Comment (kpudlik) [inline modules/currency-l0/src/main/scala/org/tessellation/currency/l0/cli/method.scala:200]: I see it repeated many times. How about doing the following?

```scala
  val lastKryoHashOrdinalOpts: Opts[Option[SnapshotOrdinal]] =
    (
      AppEnvironment.opts,
      Opts
        .argument[NonNegLong]("lastKryoHashOrdinal")
        .map(SnapshotOrdinal(_))
        .orNone
    ).mapN {
      case (env, lastKryoHash) =>
        if (env === AppEnvironment.Dev) lastKryoHash else lastKryoHashOrdinal.get(env).orElse(SnapshotOrdinal.MinValue.some)
    }
```

**Review by kpudlik — CHANGES_REQUESTED** [review]

File: modules/currency-l0/src/main/scala/org/tessellation/currency/l0/cli/method.scala (hunk 3 above)
Comment (marcinwadon) [reply modules/currency-l0/src/main/scala/org/tessellation/currency/l0/cli/method.scala:200]: @kpudlik I've already fixed that once working on the configuration, so it makes no sense to fix it here again 🤔 I would have to rebase the configuration with these changes, so I can take a second look then once combining both changes. What do you think?

**Review by kpudlik — APPROVED** [review]

**Pull request merged**